File: fc-chat/demo/part3/ch3/02-2-message-system-add-connection/src/main/java/net/prostars/messagesystem/auth/WebSocketHttpSessionHandshakeInterceptor.java
package net.prostars.messagesystem.auth;

import jakarta.servlet.http.HttpSession;
import java.util.Map;
import net.prostars.messagesystem.constant.Constants;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.server.ServerHttpRequest;
import org.springframework.http.server.ServerHttpResponse;
import org.springframework.http.server.ServletServerHttpRequest;
import org.springframework.lang.NonNull;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketHandler;
import org.springframework.web.socket.server.support.HttpSessionHandshakeInterceptor;

@Component
public class WebSocketHttpSessionHandshakeInterceptor extends HttpSessionHandshakeInterceptor {

  private static final Logger log =
      LoggerFactory.getLogger(WebSocketHttpSessionHandshakeInterceptor.class);

  @Override
  public boolean beforeHandshake(
      @NonNull ServerHttpRequest request,
      @NonNull ServerHttpResponse response,
      @NonNull WebSocketHandler wsHandler,
      @NonNull Map<String, Object> attributes) {
    if (request instanceof ServletServerHttpRequest servletServerHttpRequest) {
      Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
      if (authentication == null) {
        log.warn("WebSocket handshake failed. authentication is null.");
        response.setStatusCode(HttpStatus.UNAUTHORIZED);
        return false;
      }
      HttpSession httpSession = servletServerHttpRequest.getServletRequest().getSession(false);
      if (httpSession == null) {
        log.info("WebSocket handshake failed. httpSession is null");
        response.setStatusCode(HttpStatus.UNAUTHORIZED);
        return false;
      }
      MessageUserDetails messageUserDetails = (MessageUserDetails) authentication.getPrincipal();
      attributes.put(Constants.HTTP_SESSION_ID.getValue(), httpSession.getId());
      attributes.put(Constants.USER_ID.getValue(), messageUserDetails.getUserId());
      return true;
    } else {
      log.info("WebSocket handshake failed. request is {}", request.getClass());
      response.setStatusCode(HttpStatus.BAD_REQUEST);
      return false;
    }
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/auth/WebSocketHttpSessionHandshakeInterceptor.kt
package com.yscorp.withpush.messagesystem.auth

import com.yscorp.withpush.messagesystem.constant.IdKey
import com.yscorp.withpush.messagesystem.dto.domain.UserId
import jakarta.servlet.http.HttpSession
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.server.ServerHttpRequest
import org.springframework.http.server.ServerHttpResponse
import org.springframework.http.server.ServletServerHttpRequest
import org.springframework.lang.NonNull
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketHandler
import org.springframework.web.socket.server.support.HttpSessionHandshakeInterceptor

@Component
class WebSocketHttpSessionHandshakeInterceptor : HttpSessionHandshakeInterceptor() {
    override fun beforeHandshake(
        @NonNull request: ServerHttpRequest,
        @NonNull response: ServerHttpResponse,
        @NonNull wsHandler: WebSocketHandler,
        @NonNull attributes: MutableMap<String, Any>
    ): Boolean {
        if (request is ServletServerHttpRequest) {
            val authentication: Authentication? = SecurityContextHolder.getContext().getAuthentication()
            if (authentication == null) {
                log.warn("WebSocket handshake failed. authentication is null.")
                response.setStatusCode(HttpStatus.UNAUTHORIZED)
                return false
            }
            val httpSession: HttpSession? = request.servletRequest.getSession(false)
            if (httpSession == null) {
                log.info("WebSocket handshake failed. httpSession is null")
                response.setStatusCode(HttpStatus.UNAUTHORIZED)
                return false
            }
            val messageUserDetails = authentication.principal as MessageUserDetails
            attributes[IdKey.HTTP_SESSION_ID.value] = httpSession.id
            attributes[IdKey.USER_ID.value] = UserId(messageUserDetails.userId)
            return true
        } else {
            log.info("WebSocket handshake failed. request is {}", request.javaClass)
            response.setStatusCode(HttpStatus.BAD_REQUEST)
            return false
        }
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(WebSocketHttpSessionHandshakeInterceptor::class.java)
    }
}
